### PR TITLE
chore: add ec2 ide for attendants

### PR DESCRIPTION
**Pull request by dreamorosi**
*Issue #, if available:*

*Description of changes:*

This PR adds a new stack to the workshop dedicated to creating a browser-based IDE for attendants. 

The component involves:
- an EC2 instance in a private subnet with NAT Gateway that accepts connections only from its load balancer
- an Application Load Balancer with the instance as target, and a condition that expects the presence of an header
- a CloudFront distribution with an HTTP origin that sets the expected header
- a password to access the IDE generated at deploy time stored in Secrets Manager
- a custom resource that every 5 seconds polls the health check url of the IDE with timeout of 15 minutes - this allows the stack to either complete when the IDE is available or fail if the setup is not successful

The EC2 instance installs a series of dependencies and software needed by attendants, as well as cloning the repository, installing dependencies, and running CDK bootstrap. 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: infra/lib/attendant-ide/index.ts [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    const instanceSecurityGroup = new SecurityGroup(this, 'instance-sg', {
+      vpc: defaultVpc,
+      allowAllOutbound: true,
+    });
+
+    const userData = UserData.forLinux();
+    userData.addCommands(
+      // Install general dependencies
+      'yum install -y git docker jq zsh util-linux-user gcc make gcc-c++ libunwind java-17-amazon-corretto-headless unzip zip',
+      // Setup docker
```
[the remark this answers is not in the document]
Comment (dreamorosi) [reply infra/lib/attendant-ide/index.ts:31]: Thanks!

File: infra/lib/attendant-ide/index.ts [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+      "su - ec2-user -c 'curl -fsSL https://fnm.vercel.app/install | zsh'",
+      // Setup Python
+      `su - ec2-user -c 'export PATH="/home/ec2-user/.pyenv/bin:$PATH" && eval "$(pyenv init --path)" && pyenv install 3.11.0'`,
+      `su - ec2-user -c 'export PATH="/home/ec2-user/.pyenv/bin:$PATH" && eval "$(pyenv init --path)" && pyenv global 3.11.0'`,
+      // Setup Node
+      `su - ec2-user -c 'export PATH="/home/ec2-user/.local/share/fnm:$PATH" && eval "\`fnm env\`" && fnm install v18'`,
+      `su - ec2-user -c 'exso port PATH="/home/ec2-user/.local/share/fnm:$PATH" && eval "\`fnm env\`" && fnm default v18'`,
+      `su - ec2-user -c 'export PATH="/home/ec2-user/.local/share/fnm:$PATH" && eval "\`fnm env\`" && fnm use v18'`,
+      // Install .NET
+      "su - ec2-user -c 'curl -fsSL https://dot.net/v1/dotnet-install.sh | zsh -s -c 6.0'",
```
Comment (leandrodamascena) [inline infra/lib/attendant-ide/index.ts:51]:
```suggestion
      "rpm -Uvh https://packages.microsoft.com/config/centos/7/packages-microsoft-prod.rpm",
      "yum install -y aspnetcore-runtime-6.0 dotnet-sdk-6.0",
      "echo 'export DOTNET_ROOT=/home/ec2-user/.dotnet'                  >> /home/ec2-user/.bash_profile",
      "echo 'export PATH=\"$DOTNET_ROOT:$DOTNET_ROOT/tools:$PATH\"'        >> /home/ec2-user/.bash_profile"
```

File: infra/lib/attendant-ide/index.ts [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+      `su - ec2-user -c 'export PATH="/home/ec2-user/.local/share/fnm:$PATH" && eval "\`fnm env\`" && fnm use v18'`,
+      // Install .NET
+      "rpm -Uvh https://packages.microsoft.com/config/centos/7/packages-microsoft-prod.rpm",
+      "yum install -y aspnetcore-runtime-6.0 dotnet-sdk-6.0",
+      "echo 'export DOTNET_ROOT=/home/ec2-user/.dotnet'                  >> /home/ec2-user/.bash_profile",
+      "echo 'export PATH=\"$DOTNET_ROOT:$DOTNET_ROOT/tools:$PATH\"'        >> /home/ec2-user/.bash_profile"
+      // Install Java
+      'wget http://repos.fedorapeople.org/repos/dchen/apache-maven/epel-apache-maven.repo -O /etc/yum.repos.d/epel-apache-maven.repo',
+      'sed -i s/$releasever/6/g /etc/yum.repos.d/epel-apache-maven.repo',
+      'yum install -y apache-maven',
```
Comment (leandrodamascena) [inline infra/lib/attendant-ide/index.ts:58]:
```suggestion
      'wget https://dlcdn.apache.org/maven/maven-3/3.9.5/binaries/apache-maven-3.9.5-bin.zip',
      'unzip apache-maven-3.9.5-bin.zip -d /opt',
      'ln -s /opt/apache-maven-3.9.5/ /opt/maven',
      'echo "export M2_HOME=/opt/maven" | tee /etc/profile.d/mvn.sh',
      'echo "export PATH=${M2_HOME}/bin:${PATH}" | tee -a /etc/profile.d/mvn.sh',
      'chmod a+x /etc/profile.d/mvn.sh'
```

File: infra/lib/attendant-ide/index.ts [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+      "su - ec2-user -c 'curl https://awscli.amazonaws.com/awscli-exe-linux-aarch64.zip -o $HOME/awscliv2.zip'",
+      "su - ec2-user -c 'unzip $HOME/awscliv2.zip'",
+      "su - ec2-user -c 'sudo $HOME/aws/install'",
+      "su - ec2-user -c 'rm -rf $HOME/awscliv2.zip $HOME/aws'",
+      // Configure AWS CLI
+      `su - ec2-user -c 'aws configure set region ${Stack.of(this).region}'`,
+      // Configure git
+      `su - ec2-user -c 'git config --global init.defaultBranch main'`,
+      // Install VSCode
+      "su - ec2-user -c 'curl -fsSL https://code-server.dev/install.sh | sh'"
```
Comment (leandrodamascena) [inline infra/lib/attendant-ide/index.ts:72]:
```suggestion
      "su - ec2-user -c 'curl -fsSL https://code-server.dev/install.sh | sh'",
      "sudo systemctl enable --now code-server@ec2-user",
      "sed -i 's/127.0.0.1/0.0.0.0/g' /home/ec2-user/.config/code-server/config.yaml"

```

File: infra/lib/attendant-ide/index.ts [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+      'chmod a+x /etc/profile.d/mvn.sh'
+      // Install AWS CLI
+      "su - ec2-user -c 'curl https://awscli.amazonaws.com/awscli-exe-linux-aarch64.zip -o $HOME/awscliv2.zip'",
+      "su - ec2-user -c 'unzip $HOME/awscliv2.zip'",
+      "su - ec2-user -c 'sudo $HOME/aws/install'",
+      "su - ec2-user -c 'rm -rf $HOME/awscliv2.zip $HOME/aws'",
+      // Configure AWS CLI
+      `su - ec2-user -c 'aws configure set region ${Stack.of(this).region}'`,
+      // Configure git
+      `su - ec2-user -c 'git config --global init.defaultBranch main'`,
```
Comment (leandrodamascena) [inline infra/lib/attendant-ide/index.ts:70]:
```suggestion
      `su - ec2-user -c 'git config --global init.defaultBranch main'`,
      "su - ec2-user aws ssm put-parameter --name \"/codeserver/key\" --type "String" --value \"`uuidgen`\""
```

**Comment by dreamorosi** [comment]
I have verified that the setup of the EC2 instance works and the sum parameter is populated as expected.

I was also able to setup the Load Balancer with conditions and the Cloudfront distribution in front of it.

I am able to login into the IDE but I then get connection errors, so something is missing.

**Comment by leandrodamascena** [comment]
Let me access the account and check it!

**Comment by leandrodamascena** [comment]
Hey @dreamorosi! We have problems with the cache policy when try to Upgrade protocol to WebSocket.

Please make sure we have the following headers in the Cache-Policy and we are safe:

Sec-Websocket-Extensions
Accept
Access-Control-Request-Method
Access-Control-Request-Headers
Sec-Websocket-Key
Host
Accept-Encoding
Sec-WebSocket-Protocol
Sec-Websocket-Version

**Comment by dreamorosi** [comment]
Thanks Leo, I'll update and redeploy tomorrow morning

File: infra/lib/attendant-ide/index.ts [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+      // Setup Python
+      `su - ec2-user -c 'export PATH="/home/ec2-user/.pyenv/bin:$PATH" && eval "$(pyenv init --path)" && pyenv install 3.11.0'`,
+      `su - ec2-user -c 'export PATH="/home/ec2-user/.pyenv/bin:$PATH" && eval "$(pyenv init --path)" && pyenv global 3.11.0'`,
+      // Setup Node
+      `su - ec2-user -c 'export PATH="/home/ec2-user/.local/share/fnm:$PATH" && eval "\`fnm env\`" && fnm install v18'`,
+      `su - ec2-user -c 'exso port PATH="/home/ec2-user/.local/share/fnm:$PATH" && eval "\`fnm env\`" && fnm default v18'`,
+      `su - ec2-user -c 'export PATH="/home/ec2-user/.local/share/fnm:$PATH" && eval "\`fnm env\`" && fnm use v18'`,
+      // Install .NET
+      'rpm -Uvh https://packages.microsoft.com/config/centos/7/packages-microsoft-prod.rpm',
+      'yum install -y aspnetcore-runtime-6.0 dotnet-sdk-6.0',
```
Comment (hjgraca) [inline infra/lib/attendant-ide/index.ts:117]: After installing .NET we can also add Lambda dotnet tools to the cli.

`dotnet tool install -g Amazon.Lambda.Tools`
`dotnet new -i "Amazon.Lambda.Templates::*"`

More info: https://github.com/aws/aws-lambda-dotnet#amazonlambdatools 

Package adds commands to the dotnet cli that can be used to manage Lambda functions including deploying a function from the dotnet cli. For more information see the [README.md](https://github.com/aws/aws-lambda-dotnet/blob/master/Libraries/src/Amazon.Lambda.Tools/README.md) file for Amazon.Lambda.Tools.

Comment (dreamorosi) [reply infra/lib/attendant-ide/index.ts:117]: Added, thanks!

**Pull request merged**